src/transformations/tinshift_gpkg.cpp

```diff
@@ -0,0 +1,881 @@
+
+{
+    unsigned char *data = static_cast<unsigned char *>(dataIn);
+    for (size_t word = 0; word < word_count; word++) {
+        for (size_t i = 0; i < word_size / 2; i++) {
+            unsigned char t;
+
+            t = data[i];
+            data[i] = data[word_size - i - 1];
+            data[word_size - i - 1] = t;
+        }
+
+        data += word_size;
+    }
+}
+
+// ---------------------------------------------------------------------------
+
+static inline double sqr(double x) { return x * x; }
+static inline double squared_distance(double x1, double y1, double x2,
+                                      double y2) {
+    return sqr(x1 - x2) + sqr(y1 - y2);
+}
+static double sq_distance_point_segment(double x, double y, double x1,
+                                        double y1, double x2, double y2,
+                                        double dist12) {
+    // squared distance of point x/y to line segment x1/y1 -- x2/y2
+    const double t = ((x - x1) * (x2 - x1) + (y - y1) * (y2 - y1)) / dist12;
+    if (t <= 0.0) {
+        // closest to x1/y1
+        return squared_distance(x, y, x1, y1);
+    }
+    if (t >= 1.0) {
+        // closest to y2/y2
+        return squared_distance(x, y, x2, y2);
+    }
+
+    // closest to line segment x1/y1 -- x2/y2
+    return squared_distance(x, y, x1 + t * (x2 - x1), y1 + t * (y2 - y1));
+}
+
+// ---------------------------------------------------------------------------
+
+bool TINShiftGeopackageFile::findTriangle(
+    double x, double y, bool forwardDirection, double &lambda1, double &lambda2,
+    double &lambda3, std::vector<double> &valsVertex1,
+    std::vector<double> &valsVertex2, std::vector<double> &valsVertex3) {
+
+    sqlite3_reset(select_stmt_);
+    const bool useForwardDirection =
+        forwardDirection || !horizontalTransformation_;
+
+    // Compute barycentric coefficients
+    const auto computeLambdas = [x, y, &lambda1, &lambda2,
+                                 &lambda3](const std::array<double, 3> &vX,
+                                           const std::array<double, 3> &vY,
+                                           bool checkWithinTriangle) {
+        constexpr double EPSILON = 1e-10;
+        const double det_T = (vY[1] - vY[2]) * (vX[0] - vX[2]) +
+                             (vX[2] - vX[1]) * (vY[0] - vY[2]);
+        if (std::fabs(det_T) < EPSILON) {
+            // Degenerate triangle
+            return false;
+        }
+        lambda1 =
+            ((vY[1] - vY[2]) * (x - vX[2]) + (vX[2] - vX[1]) * (y - vY[2])) /
+            det_T;
+        lambda2 =
+            ((vY[2] - vY[0]) * (x - vX[2]) + (vX[0] - vX[2]) * (y - vY[2])) /
+            det_T;
+        if (!checkWithinTriangle) {
+            lambda3 = 1 - lambda1 - lambda2;
+            return true;
+        }
+        if (lambda1 >= -EPSILON && lambda1 <= 1 + EPSILON &&
+            lambda2 >= -EPSILON && lambda2 <= 1 + EPSILON) {
+            lambda3 = 1 - lambda1 - lambda2;
```

Comment:
Why writing lambda3 only in that case? It is strange to fill lambda1 and 2 always, but lambda3 only if return is true. I know that it can be computed (as it is there). The extra computation time is negligible.